engine/execution/ingestion/stop_at_height.go

```diff
@@ -0,0 +1,167 @@
+	return &StopControl{
+		paused:    paused,
+		commenced: paused,
+		log:       log,
+	}
+}
+
+// IsPaused returns true is block execution has been paused
+func (s *StopControl) IsPaused() bool {
+	s.RLock()
+	defer s.RUnlock()
+	return s.paused
+}
+
+// SetStopHeight sets new stop height and crash mode, and return old values:
+//   - set, whether values were previously set
+//   - height
+//   - crash
+//
+// Returns error is the stopping process has already commenced, new values will be rejected.
+func (s *StopControl) SetStopHeight(height uint64, crash bool) (bool, uint64, bool, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	oldSet := s.set
+	oldHeight := s.height
+	oldCrash := s.crash
+
+	if s.commenced {
+		return oldSet, oldHeight, oldCrash, fmt.Errorf("cannot update stop height, stopping already in progress for height %d with crash=%t", oldHeight, oldCrash)
+	}
+
+	s.set = true
+	s.height = height
+	s.crash = crash
+	s.stopAfterExecuting = flow.ZeroID
+
+	return oldSet, oldHeight, oldCrash, nil
+}
+
+// GetStopHeight returns:
+//   - set, whether values are set
+//   - height
+//   - crash
+func (s *StopControl) GetStopHeight() (bool, uint64, bool) {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.set, s.height, s.crash
+}
+
+// BlockProcessable should be called when new block is processable.
+// It returns boolean indicating if processing should skip this block.
+func (s *StopControl) BlockProcessable(b *flow.Header) bool {
+
+	s.Lock()
+	defer s.Unlock()
+
+	if !s.set {
+		return false
+	}
+
+	if s.paused {
+		return true
+	}
+
+	// skips blocks at or above requested stop height
+	if b.Height >= s.height {
+		s.log.Warn().Msgf("Skipping execution of %s at height %d because stop has been requested at height %d", b.ID(), b.Height, s.height)
+		s.commenced = true
+		return true
+	}
+
+	return false
+}
+
+// BlockFinalized should be called when a block is marked as finalized
+func (s *StopControl) BlockFinalized(ctx context.Context, execState state.ReadOnlyExecutionState, h *flow.Header) {
+
+	s.Lock()
+	defer s.Unlock()
+
+	if s.paused || !s.set {
+		return
+	}
+
+	// Once finalization reached stop height we can be sure no other fork will be valid at this height,
+	// if this block's parent has been executed, we are safe to stop or crash.
+	// This will happen during normal execution, where blocks are executed before they are finalized.
+	// However, it is possible that EN block computation progress can fall behind. In this case,
+	// we want to crash only after the execution reached the stop height.
+	if h.Height == s.height {
+
+		executed, err := state.IsBlockExecuted(ctx, execState, h.ParentID)
+		if err != nil {
+			// any error here would indicate unexpected storage error, so we crash the node
+			s.log.Fatal().Err(err).Str("block_id", h.ID().String()).Msg("failed to check if the block has been executed")
+			return
+		}
+
+		if executed {
+			s.stopExecution()
+		} else {
+			s.stopAfterExecuting = h.ParentID
+			s.set = true
```

Comment:
why updating the `set` here?

If was not set, then s.height should not have a value for `h.Height == s.height` to be true, right?

Reply:
we change internal state here (by setting stopAfterExecuting) so we mark it as set regardless

Reply:
This is a bit confusing. because set is actually `setStopHeightAndCrashed` as indicated in the comments, rather than `setStopAfterExecuting`. 

At this point `set` must have been true, if not.